content/post/moneyball_openML.Rmd

```diff
@@ -0,0 +1,361 @@
+
+## Evaluation & Resampling
+
+While the prediction indicated that the model is doing what it is supposed to, we want to have a more systematic understanding of the model performance.
+That means we want to know by how much our model is away from the truth on average.
+Cross validation investigates this question.
+In `mlr3` ten fold cross validation is constructed by the following code:
+
+```{r, size = "tiny"}
+cv10 = rsmp("cv", folds = 10)
+r = resample(task, lrn, cv10)
+```
+
+We choose some of the performance measures provided by:
+
+```{r, size = "tiny"}
+mlr_measures
+```
+
+We choose the mean absolute error (`regr.mae`) and the mean aqured error (`regr.mse`).
+
+```{r}
+scores = r$score(msrs(c("regr.mae", "regr.mse")))
+scores
+```
+
+We can also compute now by how much our model was on average wrong when predicting the runs scored.
+
+```{r}
+mean(scores$regr.mae)
+```
+
+That seems not too bad.
+Considering that on average approx. 715 runs per team per season have been scored.
+
+```{r}
+mean(data$RS)
+```
+
+## Performance boost of imputation
+
+To assess if imputation was actually beneficial we can compare our current learner with a learner which ignores the missing features.
+Normally, one would set up a benchmark for this.
+However, we want to keep things short in this use case.
+Thus, we only set up the alternative learner (with identical hyperparameters) and compare the 10-fold-cross-validated mean absolute error.
+
+We can easily ignore features by a specifc `PipeOp`
+As we are mostly interested in the numeric imputation we leave the remaining graph as it is.
+
+```{r}
+nop = po("nop")
+imp_fct = po("imputenewlvl", param_vals = list(affect_columns = selector_type("factor")))
+graph2 = po("copy", 2) %>>% gunion(list(nop, imp_fct)) %>>% po("featureunion") 
+```
+
+We then add the `PipeOpSelect` `PipeOp` to the pipeline.
+
+```{r}
+feature_names = colnames(data)[apply(!is.na(data), 2, all)]
+feature_names = c(feature_names[feature_names %in% task$feature_names], 
+                  "RankSeason", "RankPlayoffs")
+na_select = po("select")
+na_select$param_set$values$selector = selector_name(feature_names)
+graph2 = graph2 %>>% na_select
+graph2$plot(html = TRUE)
+```
+
+Now we complete the learner and apply resampling as before.
+
+```{r}
+lrn2 = GraphLearner$new(graph = graph2 %>>% polrn)
+r2 = resample(task, lrn2, cv10)
+scores2 = r2$score(msrs(c("regr.mae", "regr.mse")))
+mean(scores2$regr.mae)
+```
+
+Surprisingly, the perfomance seems to be approximately the same.
+That means that the imputed features seem not very helpful.
+We can use the `variable.importance` of the random forest.
+
+```{r}
+sort(lrn$model$regr.ranger$model$variable.importance)
+```
+
+We see that according to this the left out `OOBP` and `OSLG` seem to have only little explanatory power.
+This may be because there were simply too many instances or because the features are themselves not very powerful.
```

Comment:
Is this really helpful here? Talking about VarImp in three sentences at the end of an imputation use case? I think I would leave that out

Reply:
I tend to leave it in as it is just a quick indicator that it does not always help to impute